src/hotspot/share/cds/archiveBuilder.cpp

```diff
@@ -717,6 +717,13 @@ void ArchiveBuilder::make_klasses_shareable() {
     const char* generated = "";
     Klass* k = klasses()->at(i);
     k->remove_java_mirror();
+#ifdef _LP64
+    if (UseCompactObjectHeaders) {
+      Klass* requested_k = to_requested(k);
+      narrowKlass nk = CompressedKlassPointers::encode_not_null(requested_k, _requested_static_archive_bottom);
+      k->set_prototype_header(markWord::prototype().set_narrow_klass(nk));
+    }
+#endif //_LP64
```

Comment:
If CDS is turned off for UseCompactObjectHeaders, I don't understand this change or the one to archiveHeapWriter.  -Xshare:dump objects would be the wrong size.  If CDS is not supported, then there should be something in arguments.cpp that gives an error for that.  And write a test for that error of mixing and matching.

Reply:
Yeah, we do have code in arguments.cpp that turns off CDS if the wrong setting is used (i.e. the opposite of the default setting). If you hard-code UseCompactObjectHeaders to be true, then the archives will be written in the compact layout, and can be read. That's what the changes in share/cds implement. (Note: I regularily hard-patch the UseCompactObjectHeaders flag to be true for testing, because that also catches all the @flagless tests, and I know that Daniel does that, too.)
We disabled CDS jtreg tests when passing +UseCompactObjectHeaders via cmd line, because we would see a lot of test failures because of archive format mismatch. I am not sure if that's a good way to deal with that.

Reply:
You should add a check similar to this:

```
bool FileMapHeader::validate() {
  if (_obj_alignment != ObjectAlignmentInBytes) {
    log_info(cds)("The shared archive file's ObjectAlignmentInBytes of %d"
                  " does not equal the current ObjectAlignmentInBytes of %d.",
                  _obj_alignment, ObjectAlignmentInBytes);
    return false;
  }
```

Then, if the user wants to create a CDS archive that uses UseCompactObjectHeaders, they can do:

```
java -Xshare:dump -XX:+UseCompactObjectHeaders
java -Xshare:on -XX:+UseCompactObjectHeaders -cp . MyApp
```

Also, please create a new test case similar to this existing one:

- test/hotspot/jtreg/runtime/cds/CdsDifferentObjectAlignment.java

The checks you added in arguments.cpp should be removed.

Reply:
> You should add a check similar to this:
> 
> ```
> bool FileMapHeader::validate() {
>   if (_obj_alignment != ObjectAlignmentInBytes) {
>     log_info(cds)("The shared archive file's ObjectAlignmentInBytes of %d"
>                   " does not equal the current ObjectAlignmentInBytes of %d.",
>                   _obj_alignment, ObjectAlignmentInBytes);
>     return false;
>   }
> ```
> 
> Then, if the user wants to create a CDS archive that uses UseCompactObjectHeaders, they can do:
> 
> ```
> java -Xshare:dump -XX:+UseCompactObjectHeaders
> java -Xshare:on -XX:+UseCompactObjectHeaders -cp . MyApp
> ```
> 
> Also, please create a new test case similar to this existing one:
> 
> * test/hotspot/jtreg/runtime/cds/CdsDifferentObjectAlignment.java
> 
> The checks you added in arguments.cpp should be removed.

Thanks, Ioi, this is exactly what I've been looking for!
What about the tests-disabler, would that also not be needed anymore? It's been useful to disable CDS tests when somebody runs this with -XX:+UseCompactObjectHeaders on the cmd line.
https://github.com/rkennke/jdk/commit/e6607312b2c7e77ccd03d4d933b07b2e935d53ba

Reply:
The CDS tests generate their own archive files. The VM options specified by JTREG are included in front of the command-line for running "java -Xshare:dump", etc. So in most cases, thee options are applied equally to both dump time and run time. This means that the CDS tests will not fail with differences in ObjectAlignmentInBytes, etc.

However, some tests may be more sensitive to the JTREG VM options. These tests may need to be fixed, or tagged with `@requires vm.flagless`.

I tested with `-vmoptions:-XX:ObjectAlignmentInBytes=16` and only 3 tests failed. We'll try to fix them in [JDK-8308034](https://bugs.openjdk.org/browse/JDK-8308034)

My suggestion is to run the CDS tests with -XX:+UseCompactObjectHeaders in the command-line (we will probably add such a configuration in our CI). This will ensure better coverage. I think most tests will pass, and we can deal with the ones that might fail.